### DISCUSSION TITLE
Is it possible to do feature selection from feature importance found by the pulearn model?

### DISCUSSION THREAD

**Discussion by hlnicholls** [pulearn/pulearn#31] [Q&A]
I've been using pulearn to run it's bagging estimator using extreme gradient boosting with xgboost. My question is, is it possible to take the feature importance being found by xgboost running within the bagging estimator to be able to use that for feature selection?

My problem is that how I would normally find xgboost's feature importances is not accessible with what I'm trying:

```
xgb = XGBClassifier(use_label_encoder=False, objective='binary:logistic', eval_metric='logloss')
pu_estimator = BaggingPuClassifier(
    base_estimator = xgb, n_estimators=15)
pu_estimator.fit(X, Y)

pyplot.bar(range(len(pu_estimator.feature_importances_)), pu_estimator.feature_importances_)
pyplot.show()

AttributeError: 'BaggingPuClassifier' object has no attribute 'feature_importances_'
```

Is there a way for me to call on specifically the feature_importances_ that will be within the base estimator within the pu_estimator? Sorry if maybe I'm just missing something simple, I'm new to semi-supervised learning, and thanks for providing your package it's been very intuitive to use!

**Comment by shaypal5** [comment]
The `BaggingPUClassifier` class, which is heavily based on `scikit-learn`'s `BaggingClassifier` class, does not provide a way to compute feature importance.

You have two options at your disposal:
1. **If your base learners expose feature importance measures**, like `XGBClassifier` does, you can iterate over all base learners with ``, get the feature importance from each, and average them to get an average-based feature importance measure for the bagging ensemble model itself:
```python
feature_importances_sum = np.zeros_like(pu_estimator.estimators_[0].feature_importances_)
for estimator in pu_estimator.estimators_:
    feature_importances_sum += estimator.feature_importances_
average_feature_importances = feature_importances_sum / len(pu_estimator.estimators_)
```
2. Look at model-independent feature importance estimation methods, like here:
[https://github.com/scikit-learn/scikit-learn/issues/8898](https://github.com/scikit-learn/scikit-learn/issues/8898)
